tests/net/conn_mgr_nsos/src/main.c

```diff
@@ -0,0 +1,159 @@
+{
+	struct net_if *iface = net_if_get_default();
+	k_timeout_t conn_delay;
+	size_t optlen;
+
+	/* Default delay is 1 second */
+	optlen = sizeof(conn_delay);
+	zassert_equal(0, conn_mgr_if_get_opt(iface, 0, &conn_delay, &optlen));
+	zassert_equal(sizeof(conn_delay), optlen);
+	zassert_true(K_TIMEOUT_EQ(K_SECONDS(1), conn_delay));
+
+	/* Delay can be updated */
+	conn_delay = K_SECONDS(2);
+	zassert_equal(0, conn_mgr_if_set_opt(iface, 0, &conn_delay, sizeof(conn_delay)));
+	optlen = sizeof(conn_delay);
+	zassert_equal(0, conn_mgr_if_get_opt(iface, 0, &conn_delay, &optlen));
+	zassert_equal(sizeof(conn_delay), optlen);
+	zassert_true(K_TIMEOUT_EQ(K_SECONDS(2), conn_delay));
+
+	/* Reset to 1 second */
+	conn_delay = K_SECONDS(1);
+	zassert_equal(0, conn_mgr_if_set_opt(iface, 0, &conn_delay, sizeof(conn_delay)));
+}
+
+ZTEST(conn_mgr_nsos, test_conn_mgr_nsos_opt_error)
+{
+	struct net_if *iface = net_if_get_default();
+	k_timeout_t conn_delay = K_SECONDS(1);
+	size_t optlen;
+
+	/* Bad option */
+	optlen = sizeof(conn_delay);
+	zassert_equal(-EINVAL, conn_mgr_if_get_opt(iface, 1, &conn_delay, &optlen));
+	zassert_equal(-EINVAL, conn_mgr_if_set_opt(iface, 1, &conn_delay, sizeof(conn_delay)));
+
+	/* Bad output size */
+	optlen = sizeof(conn_delay) - 1;
+	zassert_equal(-EINVAL, conn_mgr_if_get_opt(iface, 0, &conn_delay, &optlen));
+	zassert_equal(-EINVAL, conn_mgr_if_set_opt(iface, 0, &conn_delay, sizeof(conn_delay) - 1));
+}
+
+ZTEST(conn_mgr_nsos, test_conn_mgr_nsos)
+{
+	struct net_if *iface = net_if_get_default();
+	k_timeout_t conn_delay, conn_delay_default;
+	size_t optlen = sizeof(conn_delay_default);
+
+	/* Store default delay */
+	zassert_equal(0, conn_mgr_if_get_opt(iface, 0, &conn_delay_default, &optlen));
+
+	/* Not connecting by default */
+	zassert_equal(-EAGAIN, k_sem_take(&l4_connected, K_SECONDS(2)));
+
+	/* Trigger the connection */
+	zassert_equal(0, conn_mgr_if_connect(iface));
+
+	/* Default time to connection is 1 second */
+	zassert_equal(-EAGAIN, k_sem_take(&l4_connected, K_MSEC(950)));
+	zassert_equal(0, k_sem_take(&l4_connected, K_MSEC(100)));
+	zassert_true(net_if_is_up(iface));
+
+	/* Small sleep to allow for network stack to return to idle */
+	k_sleep(K_MSEC(500));
+
+	/* Disconnect (actioned immediately) */
+	zassert_equal(0, conn_mgr_if_disconnect(iface));
+	zassert_equal(0, k_sem_take(&l4_disconnected, K_MSEC(100)));
+	zassert_false(net_if_is_up(iface));
+
+	/* Try again with custom timeout */
+	conn_delay = K_MSEC(500);
+	zassert_equal(0, conn_mgr_if_set_opt(iface, 0, &conn_delay, sizeof(conn_delay)));
+
+	/* Trigger the connection */
+	zassert_equal(0, conn_mgr_if_connect(iface));
+
+	/* Should connect after 500ms this time */
+	zassert_equal(-EAGAIN, k_sem_take(&l4_connected, K_MSEC(450)));
+	zassert_equal(0, k_sem_take(&l4_connected, K_MSEC(100)));
+	zassert_true(net_if_is_up(iface));
+
+	/* Small sleep to allow for network stack to return to idle */
+	k_sleep(K_MSEC(500));
```

Comment:
I just wonder if we could make the delays shorter or remove them, I am counting several second delay when this test is run. This might not sound much but these things add up when thousands of tests are run. We have been trying to get rid of these kind of sleeps in other parts of the networking tests.

Reply:
Indeed, my bad, no worries then.